docs/developer/architecture-and-design/workflow-design.md

```diff
@@ -0,0 +1,228 @@
+Parameters are identified via their type, i.e., we will require defining a domain-specific type for each parameter, such as `WavelengthBinning`.
+
+For nested workflows, we can use a child injector, which provides a scope for parameters.
+Parent-scopes can be searched for parameters that are not found in the child-scope, providing a mechanism for "global" parameters.
+
+*Note:
+The idea of using child injectors was discarded during implementation for a number of reasons.
+Parameters in a "nested" scope can now be realized using the mechanism of generic providers.*
+
+### Metadata handling
+
+There have been a number of discussions around metadata handling.
+For example, the support (or non-support) of an arbitrary `attrs` dict as part of `scipp.Variable` and `scipp.DataArray`.
+Furthermore, we may have metadata that is part of the data-catalog, which may partially overlap with the metadata that is part of the data itself.
+The current conclusion is that any attempt to handle metadata in a generic and automatic way will not work.
+Therefore, if a user wants to provide metadata for a workflow result, they must do so *explicitly* by specifying functions that can assemble that metadata.
+As with regular results, this can be done by injecting the input metadata into the function that computes the result's metadata.
+
+### Domain-specific types
+
+The system will use domain-specific types to identify workflow steps and parameters.
+The system does not require subclassing or use of decorators.
+Domain-specific types can be defined as regular classes or subclasses of existing types.
+In many cases, a simple alias will be sufficient.
+
+`typing.NewType` can be used as a simple way of creating a domain-specific type alias for type-checking.
+For example, we can use it to create a type for a `scipp.DataArray` that represents a transmission monitor.
+This avoids a more complex solution such as creating a wrapper class or a subclass:
+
+```python
+import typing
+
+TransmissionMonitor = typing.NewType('TransmissionMonitor', scipp.DataArray)
+```
+
+Note that this does not create a new type, but rather a new name for an existing type.
+That is, `isinstance(monitor, TransmissionMonitor)` does not work, since `TransmissionMonitor` is not a type.
+Furthermore, operations will revert to the underlying type, e.g., `monitor * 2` will return a `scipp.DataArray`.
+For this application this would actually be desired behavior:
+Applying an operation to a domain type will generally result in a different type, so falling back to the underlying type is the correct behavior and forces the user to be explicit about the type of the result.
+
+## Notes
+
+- Earlier versions of this design document included a detailed discussion on how to handle nested workflows using child-injectors.
+ During initial implementation efforts this idea was discarded, as it was found to be too complicated and not necessary.
```

Comment:
Duplicate of l. 181f.

Reply:
Um, what is duplicate?

Reply:
Maybe not a full duplicate. But the other section also mentions that nested workflows were considered but not implemented.

Reply:
Isn't the section talking about nested parameters, as opposed to nested workflows?